python/mxnet/runtime.py

```diff
@@ -0,0 +1,82 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# coding: utf-8
+# pylint: disable=not-an-iterable
+
+"""runtime detection of compile time features in the native library"""
+
+import ctypes
+import enum
+from .base import _LIB, check_call, mx_uint, py_str
+
+
+def _feature_names_available():
+    """
+
+    :return:
+    """
+    feature_list = ctypes.POINTER(ctypes.c_char_p)()
+    feature_list_sz = ctypes.c_size_t()
+    check_call(_LIB.MXRuntimeFeatureList(ctypes.byref(feature_list_sz), ctypes.byref(feature_list)))
+    feature_names = []
+    for i in range(feature_list_sz.value):
+        feature_names.append(py_str(feature_list[i]))
+    return feature_names
+
+Feature = enum.Enum('Feature', {name: index for index, name in enumerate(_feature_names_available())})
+
+def features_available():
```

Comment:
a single api that lists all features and their values seems easier to maintain and it wouldn't compromise the usability.

Reply:
See comment above, there's features available and enabled. These are two sets where enabled is a subset of available.

Reply:
What's wrong with having a single method to return the complete list of features in string, the mask number, and a binary state indicator for each feature on whether a feature is enabled or not?

Reply:
ah ok I understood you wanted too different API calls. That sounds reasonable to me.

Reply:
To return a list of features, mask number etc, that would be array of struct. We would need either another call to first get the number of features so we can reserve an array of struct of the propper size in python using ctypes. Another option is that this array of struct is initialized statically somewhere like LibraryInitializer or as a singleton, which is always a bit of an anti pattern for testing / mocking, or asuming there's only one instance of somthing. Third option is on the heap but that opens the door for leaks. What would in your opinion be the best choice?

Reply:
I think having a call that gets the number first is fine. This should make it an idempotent function.